modules/siddhi-core/src/main/java/io/siddhi/core/table/record/AbstractQueryableRecordTable.java

```diff
@@ -49,28 +70,525 @@
+            updateConditionParameterMaps.add(variableMap);
+
+            Map<String, Object> variableMapForUpdateSet = new HashMap<>();
+            for (Map.Entry<String, ExpressionExecutor> entry :
+                    recordTableCompiledUpdateSet.getExpressionExecutorMap().entrySet()) {
+                variableMapForUpdateSet.put(entry.getKey(), entry.getValue().execute(stateEvent));
+            }
+            updateSetParameterMaps.add(variableMapForUpdateSet);
+            timestamp = stateEvent.getTimestamp();
+        }
+        if (recordTableHandler != null) {
+            recordTableHandler.update(timestamp, recordStoreCompiledCondition.compiledCondition,
+                    updateConditionParameterMaps, recordTableCompiledUpdateSet.getUpdateSetMap(),
+                    updateSetParameterMaps);
+        } else {
+            update(recordStoreCompiledCondition.compiledCondition, updateConditionParameterMaps,
+                    recordTableCompiledUpdateSet.getUpdateSetMap(), updateSetParameterMaps);
+        }
+    }
+
+    @Override
+    public boolean contains(StateEvent matchingEvent, CompiledCondition compiledCondition)
+            throws ConnectionUnavailableException {
+        RecordStoreCompiledCondition recordStoreCompiledCondition;
+        if (isCacheEnabled) {
+            RecordStoreCompiledCondition compiledConditionTemp = (RecordStoreCompiledCondition) compiledCondition;
+            CompiledConditionWithCache compiledConditionWithCache = (CompiledConditionWithCache)
+                    compiledConditionTemp.compiledCondition;
+            recordStoreCompiledCondition = new RecordStoreCompiledCondition(compiledConditionTemp.
+                    variableExpressionExecutorMap, compiledConditionWithCache.getStoreCompileCondition());
+            if (cachedTable.contains(matchingEvent, compiledConditionWithCache.getCacheCompileCondition())) {
+                return true;
+            }
+        } else {
+            recordStoreCompiledCondition = ((RecordStoreCompiledCondition) compiledCondition);
+        }
+        Map<String, Object> containsConditionParameterMap = new HashMap<>();
+        for (Map.Entry<String, ExpressionExecutor> entry :
+                recordStoreCompiledCondition.variableExpressionExecutorMap.entrySet()) {
+            containsConditionParameterMap.put(entry.getKey(), entry.getValue().execute(matchingEvent));
+        }
+        if (recordTableHandler != null) {
+            return recordTableHandler.contains(matchingEvent.getTimestamp(), containsConditionParameterMap,
+                    recordStoreCompiledCondition.compiledCondition);
+        } else {
+            return contains(containsConditionParameterMap, recordStoreCompiledCondition.compiledCondition);
+        }
+    }
+
+    @Override
+    public void updateOrAdd(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
+                            CompiledCondition compiledCondition, CompiledUpdateSet compiledUpdateSet,
```

Comment:
Do we need to synchronize the cachedTable update and the table update ? in all the insert or update methods?